unstructured/ingest/v2/processes/connectors/salesforce.py

```diff
@@ -0,0 +1,291 @@
+Subject: $subject
+From: $from_email
+To: $to_email
+Content-Type: multipart/alternative; boundary="00000000000095c9b205eff92630"
+--00000000000095c9b205eff92630
+Content-Type: text/plain; charset="UTF-8"
+$textbody
+--00000000000095c9b205eff92630
+Content-Type: text/html; charset="UTF-8"
+$htmlbody
+--00000000000095c9b205eff92630--
+""",
+)
+
+
+@dataclass
+class SalesforceAccessConfig(AccessConfig):
+    consumer_key: str
+    private_key: str
+
+    @requires_dependencies(["cryptography"])
+    def get_private_key_value_and_type(self) -> Tuple[str, Type]:
+        from cryptography.hazmat.primitives import serialization
+
+        try:
+            serialization.load_pem_private_key(data=self.private_key.encode("utf-8"), password=None)
+        except ValueError:
+            pass
+        else:
+            return self.private_key, str
+
+        if Path(self.private_key).is_file():
+            return self.private_key, Path
+
+        raise ValueError("private_key does not contain PEM private key or path")
+
+
+@dataclass
+class SalesforceConnectionConfig(ConnectionConfig):
+    username: str
+    access_config: SalesforceAccessConfig = enhanced_field(sensitive=True)
+
+    @requires_dependencies(["simple_salesforce"], extras="salesforce")
+    def get_client(self) -> "Salesforce":
+        from simple_salesforce import Salesforce
+
+        pkey_value, pkey_type = self.access_config.get_private_key_value_and_type()
+
+        return Salesforce(
+            username=self.username,
+            consumer_key=self.access_config.consumer_key,
+            privatekey_file=pkey_value if pkey_type is Path else None,
+            privatekey=pkey_value if pkey_type is str else None,
+            version=SALESFORCE_API_VERSION,
+        )
+
+
+@dataclass
+class SalesforceIndexerConfig(IndexerConfig):
+    categories: list[str]
+
+
+@dataclass
+class SalesforceIndexer(Indexer):
+    connection_config: SalesforceConnectionConfig
+    index_config: SalesforceIndexerConfig
+
+    def get_file_extension(self, record_type) -> str:
+        if record_type == "EmailMessage":
+            extension = ".eml"
+        elif record_type in ["Account", "Lead", "Case", "Campaign"]:
+            extension = ".xml"
+        else:
+            raise MissingCategoryError(
+                f"There are no categories with the name: {record_type}",
+            )
+        return extension
+
+    @requires_dependencies(["simple_salesforce"], extras="salesforce")
+    def list_files(self) -> list[FileData]:
+        """Get Salesforce Ids for the records.
+        Send them to next phase where each doc gets downloaded into the
+        appropriate format for partitioning.
+        """
+        from simple_salesforce.exceptions import SalesforceMalformedRequest
+
+        client = self.connection_config.get_client()
+
+        files_list = []
+        for record_type in self.index_config.categories:
+            if record_type not in ACCEPTED_CATEGORIES:
```

Comment:
Can add this as a check in the __post_init__ method of the indexer config since that's where these get set. By the time this method runs, it can assume the categories are all valid.